lib/tokenizer/rainscriptTokenizer.js

```diff
@@ -0,0 +1,72 @@
+import Lexer from 'lex';
+import { _ } from 'underscore';
+import { rainscriptTokenizerRules } from './rainscriptTokenizerRules';
+
+const ccount = require('ccount');
+
+// the regex-based lexer for the Rainscript tokenizer
+let _lexer = new Lexer(token(null));
+
+// grab rules from tokenizerRules file
+rainscriptTokenizerRules.forEach(rule =>
+  _lexer.addRule(rule.pattern, token(rule.type), rule.start)
+);
+
+/**
+ * A factory method that creates a new Rainscript tokenizer instance.
+ * @returns an instance of Rainscript's tokenizer
+ */
+export const rainscriptTokenizer = () => {
+  // wrap lexer prototype so that tokenizers can be thread-safe
+  let lexer = Object.create(_lexer);
+
+  return {
+    /**
+     * Collects all tokens of an input text. Also collects metadata information
+     * such as the line number and column index of the parsed token.
+     * @param {String} text the text to parse tokens from
+     * @return {{type: any, lexeme: String, line: int, column: int}}
+     */
+    tokenize (text) {
+      if (!_.isString(text)) {
+        throw new TypeError('tokenize only accepts a String');
+      }
+      lexer.setInput(text);
+      let line = 1;
+      let column = 1;
+      let token;
+      let tokens = [];
+      while (!_.isUndefined(token = lexer.lex())) {
```

Comment:
My concern is, this particular line does two important actions. Also, token is only used inside the loop.

Can you try creating a generator function that gets the next `lex()` value and iterate through it? something like this
```js
function* lex (lexer) {
  while (true) {
    yield lexer.lex();
  }
}

// later
for(let token of lex(lexer)) {
  if (_.isUndefined(token)) {
    break;
  }
  // ... do stuff here
}
```

Reply:
Wow, never knew there was something like generators in ES6. But the generator action would only be solely used in one function. Any advantage of it over simply doing a `while(true)` loop?

```js
while (true) {
  let token = lexer.lex();
  if (_.isUndefined(token)) {
    break;
  }
  // ... do stuff here
}
```

Reply:
I just don't like `while (true)` and instantiating a variable outside of the block it is used in. I guess we can settle for `while (true)`, unless you have other suggestions. :smile: